src/persist-client/examples/persistent_source.rs

```diff
@@ -0,0 +1,568 @@
+        });
+
+        let _ = tokio::join!(
+            source_task,
+            prefetcher_task,
+            decoder_task,
+            timestamper_task,
+            consumer_task
+        );
+
+        Ok(())
+    }
+
+    /// Reads batches from the given `source` and forwards them to `tx`.
+    async fn read_source<S, T>(
+        mut source: S,
+        batch_size: usize,
+        tx: mpsc::Sender<Batch<S::D, S::SourceTimestamp>>,
+    ) where
+        S: Source<SourceTimestamp = T>,
+        T: timely::PartialOrder + PartialOrd + Debug,
+    {
+        loop {
+            let batch = source.read_batch(batch_size).await;
+
+            if let Some(batch) = batch {
+                if let Err(_) = tx.send(batch).await {
+                    error!("receiver dropped");
+                    return;
+                }
+            } else {
+                tokio::time::sleep(Duration::from_millis(10)).await;
+            }
+        }
+    }
+
+    /// Notifies the [`Timestamper`] of newly available source uppers, in an effort to parallelize
+    /// the minting and retrieval of new timestamp bindings and decoding of updates.
+    async fn prefetch_timestamp_bindings<D: Debug, TS, ST, AT>(
+        timestamper: TS,
+        mut rx: mpsc::Receiver<Batch<D, ST>>,
+        tx: mpsc::Sender<Batch<D, ST>>,
+    ) where
+        TS: Timestamper<SourceTimestamp = ST, AssignedTimestamp = AT> + Send + Sync + 'static,
+        ST: timely::PartialOrder + PartialOrd + Debug + Send + Sync + Copy + 'static,
+        AT: Debug + Copy,
+        D: Send,
+    {
+        while let Some(batch) = rx.recv().await {
+            let timestamper = timestamper.clone();
+            let upper = batch.upper.clone();
+            let lower = batch.lower.clone();
+            tokio::task::spawn(async move {
+                let _ = timestamper.get_bindings(lower, upper).await;
+            });
+
+            if let Err(_) = tx.send(batch).await {
+                error!("receiver dropped");
+                return;
+            }
+        }
+    }
+
+    /// A mock "decode" operation that uses [`Debug`] to turn incoming updates into strings.
+    async fn decode<D: Debug, T: Debug>(
+        mut rx: mpsc::Receiver<Batch<D, T>>,
+        tx: mpsc::Sender<Batch<String, T>>,
+    ) {
+        while let Some(batch) = rx.recv().await {
+            let mut decoded_updates = Vec::new();
+
+            for (update, ts) in batch.updates {
+                decoded_updates.push((format!("{:?}", update), ts));
+            }
+
+            let decoded_batch = Batch {
+                updates: decoded_updates,
+                lower: batch.lower,
+                upper: batch.upper,
+            };
+            if let Err(_) = tx.send(decoded_batch).await {
+                error!("receiver dropped");
+                return;
+            }
+        }
+    }
+
+    /// Assigns timestamps to updates using the given `timestamper`.
+    async fn assign_timestamp<D, TS, ST, AT>(
```

Comment:
can we use the language from the reclocking design doc for the various applicable pieces? e.g. remap, reclocked, reclocking?

Reply:
👌